uniffi/src/lib.rs

```diff
@@ -476,6 +476,66 @@ unsafe impl<V: ViaFfi> ViaFfi for HashMap<String, V> {
     }
 }
 
+use serde_json::Value;
+
+/// Support for passing JSON objects.
+///
+/// For correctness and speed of implementation, objects are serialized into a String.
+/// This can definitely be improved in efficiency, however high-quality JSON parsers and
+/// stringifiers are available on both sides of the FFI.
+///
+/// Only top-level objects are supported: i.e. the value must be a `Value::Object`.
+/// If the Rust side tries to pass back a `JSONArray` or a scalar, and the foreign language side
+/// is expecting a dictionary shaped value, it would be a failure.
+///
+/// In this implementation, the value is wrapped in an object with the error message as the key.
+unsafe impl ViaFfi for Value {
+    type FfiType = RustBuffer;
+
+    fn lower(self) -> Self::FfiType {
+        match to_supported_string(&self) {
+            Ok(string) => string.lower(),
+            Err(_) => panic!("Unserializable JSON"),
+        }
+    }
+
+    // The argument here *must* be a uniquely-owned `RustBuffer` previously obtained
+    // from `lower` above, and hence must be the bytes of a valid rust string.
+    fn try_lift(v: Self::FfiType) -> Result<Self> {
+        let json_string = String::try_lift(v)?;
+        Ok(serde_json::from_str(&json_string)?)
+    }
+
+    fn write<B: BufMut>(&self, buf: &mut B) {
+        match to_supported_string(&self) {
+            Ok(json_string) => json_string.write(buf),
+            Err(_) => panic!("Unserializable JSON"),
+        };
+    }
+
+    fn try_read<B: Buf>(buf: &mut B) -> Result<Self> {
+        let json_string = String::try_read(buf)?;
+        Ok(serde_json::from_str(&json_string)?)
+    }
+}
+
+fn to_supported_string(value: &Value) -> Result<String> {
+    match value {
+        Value::Object(_) => Ok(serde_json::to_string(value)?),
+        _ => {
+            // We can either cause a panic! when unsupported/unsanitized JSON is
+            // accidently given to us, or we can wrap it up as an object with an error message
+            // as a key. Neither is ideal, but this should cause failure down stream quick enough
+            // for the developer to catch it.
```

Comment:
sorry, but I'm not a fan of this. I think we should panic. I think there are a number of theoretical panics in even the simple type operations. Regardless though, this gets to the heart of the risks in introducing a "fallible" type (which I'm not against, but opens some interesting questions)

Reply:
Good to get a second opinion on this. I'll change it to a runtime check and panic, and add some "sanitize your inputs" warnings in the docs.